### DISCUSSION TITLE
ST7701 and SenseCap Indicator D1

### DISCUSSION THREAD

**Discussion by grafail** [lvgl-micropython/lvgl_micropython#230] [Q&A]
I have a Sensecap Indicator D1(https://wiki.seeedstudio.com/SenseCAP_Indicator_Dive_into_the_Hardware/) and I am having trouble getting one of the sample scripts to work. 
I have implemented a custom type to initialise ST7701 to match the sample repo of the device:
https://github.com/Seeed-Solution/SenseCAP_Indicator_ESP32/blob/c16c51a0d87abae1f86c6bdf209f45a3d12283d8/components/bsp/src/boards/lcd_panel_config.c#L854
I feel I am getting the Spi3Wire configuration wrong. It seems that the screen is on a PCA9535 expander. I am not sure if that is meant to be used with Spi3Wire or how to reference its internal port as SPI and I2C seem to be different.
Any help or pointers would be greatly appreciated. Currently only the backlight seems to turn on. I would love to get this working.

My current code:
```python
print("INITIALISING SCREEN")

from micropython import const # NOQA

_WIDTH = const(480)
_HEIGHT = const(480)

import lcd_bus # NOQA
# File with custom init code
from init_code import init_screen
from spi3wire import Spi3Wire # NOQA
import gc
gc.collect()


spi3wire = Spi3Wire(
sck=41,
miso=47,
cs=39,
cs_active_high=False,
)

print("SPI WIRE DONE")
display_bus = lcd_bus.RGBBus(
hsync=16,
vsync=17,
de=18,
pclk=21,
data0=4,
data1=3,
data2=2,
data3=1,
data4=0,
data5=10,
data6=9,
data7=8,
data8=7,
data9=6,
data10=5,
data11=15,
data12=14,
data13=13,
data14=12,
data15=11,
freq=13 * 1000 * 1000,
hsync_back_porch=50,
hsync_front_porch=10,
hsync_pulse_width=8,
vsync_back_porch=20,
vsync_front_porch=10,
vsync_pulse_width=8,
)

print("RGB BUS DONE")
import st7701
import lvgl as lv # NOQA

fb1 = display_bus.allocate_framebuffer(int(_WIDTH * _HEIGHT * 2 / 10), lcd_bus.MEMORY_SPIRAM)#lcd_bus.MEMORY_INTERNAL | lcd_bus.MEMORY_DMA)
fb2 = display_bus.allocate_framebuffer(int(_WIDTH * _HEIGHT * 2 / 10), lcd_bus.MEMORY_SPIRAM)#lcd_bus.MEMORY_INTERNAL | lcd_bus.MEMORY_DMA)
print("BUFFER ALLOCATED")
class CustomST7701(st7701.ST7701):

    def _spi_3wire_init(self, type):  # NOQA
        # Overriding to run my version of the init sequence
        init_screen(self)

display = CustomST7701(
data_bus=display_bus,
spi_3wire=spi3wire,
display_width=_WIDTH,
display_height=_HEIGHT,
frame_buffer1=fb1,
frame_buffer2=fb2,
backlight_pin=45,
color_space=lv.COLOR_FORMAT.RGB565,
bus_shared_pins=False
)

display.set_power(True)
display.init(1)
display.set_backlight(50)

scrn = lv.screen_active()
scrn.set_style_bg_color(lv.color_hex(0x000000), 0)

slider = lv.slider(scrn)
slider.set_size(300, 50)
slider.center()

label = lv.label(scrn)
label.set_text('HELLO WORLD!')
label.align(lv.ALIGN.CENTER, 0, -50)

import task_handler
th = task_handler.TaskHandler()
```

**Comment by kdschlosser** [comment]
what version of that display do you have?

**Reply by grafail** [reply]
D1 (version without any sensors or Lora)
https://www.seeedstudio.com/SenseCAP-Indicator-D1-p-5643.html

**Reply by kdschlosser** [reply]
There are 2 variations of that display one is an GX and the other is the DX do you know which one you have?

**Reply by grafail** [reply]
I believe it is the GX variant

**Comment by kdschlosser** [comment]
It looks like no matter the version they use the same io expander which is the tca9535. I am in the process of making modifications to add  io expanders to the firmware. I already had written a framework for it so that is already done. II have written the driver for your expander so that is also done. I also have to modify the display driver framework, the pointer framework and the 3 wire SPI so they can accept an io expander pin. The io expander framework mimics the machine.Pin framework so it is not going to be too much work to update things to be able to use it.

**Reply by grafail** [reply]
That sounds great, thank you!

**Reply by kdschlosser** [reply]
I am working on this however, it is going to be a bit before I am done. I have to make a very large code change in order to get things working correctly. You are going to have to be patient as it is quite a bit of work to get everything up to snuff so it will function properly.

**Reply by grafail** [reply]
Yeah of course. I appreciate all the help.

**Comment by grafail** [comment]
This is my current version of this file, it still seems to only turn on the backlight:

```py
print("INITIALISING SCREEN")

from micropython import const # NOQA

_WIDTH = const(480)
_HEIGHT = const(480)

import lcd_bus # NOQA
# File with custom init code
from init_code import init_screen
from spi3wire import Spi3Wire # NOQA
import gc
gc.collect()


import i2c
import tca9535

expander_bus = i2c.I2C.Bus(host=1, scl=40, sda=39)

tca9535.Pin.set_device(i2c.I2C.Device(expander_bus, tca9535.I2C_ADDR, tca9535.BITS))

lcd_cs_pin = tca9535.Pin(tca9535.EXIO4, tca9535.Pin.OUT, value=0)
lcd_rest_pin = tca9535.Pin(tca9535.EXIO5, tca9535.Pin.OUT, value=0)

spi3wire = Spi3Wire(
sck=41,
miso=48,
cs=lcd_cs_pin,
cs_active_high=False,
)

print("SPI WIRE DONE")
display_bus = lcd_bus.RGBBus(
hsync=16,
vsync=17,
de=18,
pclk=21,
data0=4,
data1=3,
data2=2,
data3=1,
data4=0,
data5=10,
data6=9,
data7=8,
data8=7,
data9=6,
data10=5,
data11=15,
data12=14,
data13=13,
data14=12,
data15=11,
freq=13 * 1000 * 1000,
hsync_back_porch=50,
hsync_front_porch=10,
hsync_pulse_width=8,
vsync_back_porch=20,
vsync_front_porch=10,
vsync_pulse_width=8,
)

print("RGB BUS DONE")
import st7701
import lvgl as lv # NOQA
sz = int(_WIDTH * _HEIGHT * 2 / 10)
fb1 = display_bus.allocate_framebuffer(sz, lcd_bus.MEMORY_SPIRAM)
fb2 = display_bus.allocate_framebuffer(sz, lcd_bus.MEMORY_SPIRAM)
print("BUFFER ALLOCATED")
class CustomST7701(st7701.ST7701):

    def _spi_3wire_init(self, type):  # NOQA
        # Overriding to run my version of the init sequence
        init_screen(self)

display = CustomST7701(
data_bus=display_bus,
spi_3wire=spi3wire,
display_width=_WIDTH,
display_height=_HEIGHT,
frame_buffer1=fb1,
frame_buffer2=fb2,
backlight_pin=45,
reset_pin=lcd_rest_pin,
color_space=lv.COLOR_FORMAT.RGB565,
bus_shared_pins=False
)

display.set_power(True)
display.init(1)
display.set_backlight(100)

import task_handler
th = task_handler.TaskHandler()

scrn = lv.screen_active()
scrn.set_style_bg_color(lv.color_hex(0x000000), 0)

slider = lv.slider(scrn)
slider.set_size(300, 50)
slider.center()

label = lv.label(scrn)
label.set_text('HELLO WORLD!')
label.align(lv.ALIGN.CENTER, 0, -50)
```

**Reply by grafail** [reply]
This is the init code I used
[init_code.py.txt](https://github.com/user-attachments/files/18298300/init_code.py.txt)

**Reply by kdschlosser** [reply]
That's right, you have the st7701 display....

I have this display 

https://www.buydisplay.com/spi-1280x400-7-84-inch-ips-tft-lcd-module-optl-capacitive-touch

which is an I8080 bus but also has the spi3wire and I have not been able to get it to run either. There is someone that has used the spi3wire driver for their display and they have not had an issue with it.

I am going to do some more digging to see if I can locate a 3wire driver for the ESP32's

**Reply by kdschlosser** [reply]
OK I just rewrote the spi3wire. It has more options now so it might work for ya. I will give it a try later today.

**Comment by grafail** [comment]
Experimented with the new spi3wire a bit, I made a few changes to resolve some compilation issues (https://github.com/grafail/lvgl_micropython/tree/esp3wire-test), but it seems I am still not getting something except backlight.

```py
print("INITIALISING SCREEN")

from micropython import const # NOQA

_WIDTH = const(480)
_HEIGHT = const(480)

import lcd_bus # NOQA
# File with custom init code
from init_code import init_screen
from spi3wire import Spi3Wire # NOQA
import gc
gc.collect()

spi3wire = Spi3Wire(
scl=40,
sda=39,
cs=4,
freq=4 * 100 * 1000,
spi_mode=3
)

print("SPI WIRE DONE")
display_bus = lcd_bus.RGBBus(
hsync=16,
vsync=17,
de=18,
pclk=21,
data0=4,
data1=3,
data2=2,
data3=1,
data4=0,
data5=10,
data6=9,
data7=8,
data8=7,
data9=6,
data10=5,
data11=15,
data12=14,
data13=13,
data14=12,
data15=11,
freq=13 * 1000 * 1000,
hsync_back_porch=50,
hsync_front_porch=10,
hsync_pulse_width=8,
vsync_back_porch=20,
vsync_front_porch=10,
vsync_pulse_width=8,
)

print("RGB BUS DONE")
import st7701
import lvgl as lv # NOQA
sz = int(_WIDTH * _HEIGHT * 2 / 10)
fb1 = display_bus.allocate_framebuffer(sz, lcd_bus.MEMORY_SPIRAM)
fb2 = display_bus.allocate_framebuffer(sz, lcd_bus.MEMORY_SPIRAM)
print("BUFFER ALLOCATED")
class CustomST7701(st7701.ST7701):

    def _spi_3wire_init(self, type):  # NOQA
        # Overriding to run my version of the init sequence
        init_screen(self)

display = CustomST7701(
data_bus=display_bus,
spi_3wire=spi3wire,
display_width=_WIDTH,
display_height=_HEIGHT,
frame_buffer1=fb1,
frame_buffer2=fb2,
backlight_pin=45,
color_space=lv.COLOR_FORMAT.RGB565,
bus_shared_pins=False
)

display.set_power(True)
display.init(1)
display.set_backlight(100)

import task_handler
th = task_handler.TaskHandler()

scrn = lv.screen_active()
scrn.set_style_bg_color(lv.color_hex(0x000000), 0)

slider = lv.slider(scrn)
slider.set_size(300, 50)
slider.center()

label = lv.label(scrn)
label.set_text('HELLO WORLD!')
label.align(lv.ALIGN.CENTER, 0, -50)
```
Feel this is fairly close to working, but it is quite hard to debug. I suspect `cs=4` may be problematic.

**Reply by kdschlosser** [reply]
you can add `--debug` to the build arguments to enable the debugging output. I have all of the compilation errors sorted out in the last commit I did..

Why do you suspect the `cs=4` as being a problem?

**Reply by kdschlosser** [reply]
I see where the issue is...

You see the parameter `bus_shared_pins` for the RGB display driver?

set that to `True`. That flag is to indicate that the display bus is sharing pins with the spi3wire. If there are shared pins then the spi3wire needs to be torn down before the display bus gets initialized but after the initialization commands are sent. That is what that flag does.

**Reply by grafail** [reply]
I changed scl, sda to the correct pin numbers, set bus_shared_pins=True and it seems to work (partly). I do get proper output now, but only after I disconnect the cable for a few seconds and then plug it back. If I do a soft reset, it seems to not work.

**Reply by kdschlosser** [reply]
soft reset is not currently supported. I am working on getting it working but it is more of an issue with MicroPython and the hardware than it is with anything else.

**Reply by kdschlosser** [reply]
I am very happy to hear that it is working now.. I did add software rotation to the RGB driver. It will slow things down but not to the point of being horrible which is what it was using LVGL's internal rotation.

**Reply by grafail** [reply]
Seems that the issue gets triggered if I soft reset too fast, seems that even works, If I wait for it to fully initiate. I tested the version on the main branch and I needed a couple small changes on top of it to get it working.
https://github.com/grafail/lvgl_micropython/commit/8f5ec2a2760acb4389cd591d76269cfd7a276420

Thank you again for helping me set this up!

**Reply by kdschlosser** [reply]
I fixed those 2 issues.

OK so the software reset is now working so long as the hardware has fully initialized... That's good to know.

**Reply by jbalague** [reply]
> I changed scl, sda to the correct pin numbers, set bus_shared_pins=True and it seems to work (partly). I do get proper output now, but only after I disconnect the cable for a few seconds and then plug it back. If I do a soft reset, it seems to not work.

Testing your code but I'm stuck at the backlight only. Can you share the correct scl and sda pin numbers?

**Reply by grafail** [reply]
@jbalague This is the latest code I have along with the part to get touch working.

```python
print("INITIALISING SCREEN")

from micropython import const # NOQA

_WIDTH = const(480)
_HEIGHT = const(480)

import lcd_bus # NOQA
# File with custom init code
from spi3wire import Spi3Wire # NOQA
import gc
gc.collect()

spi3wire = Spi3Wire(
scl=41,
sda=48,
cs=4,
freq=5 * 100 * 1000,
spi_mode=3,
)

print("SPI WIRE DONE")
display_bus = lcd_bus.RGBBus(
    hsync=16,
    vsync=17,
    de=18,
    pclk=21,
    data0=15,
    data1=14,
    data2=13,
    data3=12,
    data4=11,
    data5=10,
    data6=9,
    data7=8,
    data8=7,
    data9=6,
    data10=5,
    data11=4,
    data12=3,
    data13=2,
    data14=1,
    data15=0,

    freq=13 * 1000 * 1000,
    hsync_back_porch=50,
    hsync_front_porch=10,
    hsync_pulse_width=8,
    vsync_back_porch=20,
    vsync_front_porch=10,
    vsync_pulse_width=8,
)

print("RGB BUS DONE")
import st7701
import lvgl as lv # NOQA
sz = int(_WIDTH * _HEIGHT * 2 / 5)
fb1 = display_bus.allocate_framebuffer(sz, lcd_bus.MEMORY_SPIRAM)
fb2 = display_bus.allocate_framebuffer(sz, lcd_bus.MEMORY_SPIRAM)
print("BUFFER ALLOCATED")

display = st7701.ST7701(
data_bus=display_bus,
spi_3wire=spi3wire,
display_width=_WIDTH,
display_height=_HEIGHT,
frame_buffer1=fb1,
frame_buffer2=fb2,
backlight_pin=45,
color_space=lv.COLOR_FORMAT.RGB565,
bus_shared_pins=True
)


display.set_power(True)
display.init(17)
display.set_backlight(100)

import ft6x36  # NOQA
import i2c

i2c_bus = i2c.I2C.Bus(host=1, scl=40, sda=39, freq=5*100*1000, use_locks=False)
touch_dev = i2c.I2C.Device(bus=i2c_bus, dev_id=72, reg_bits=ft6x36.BITS)

indev = ft6x36.FT6x36(touch_dev, startup_rotation=lv.DISPLAY_ROTATION._180)

import task_handler
th = task_handler.TaskHandler()

scrn = lv.screen_active()
scrn.set_style_bg_color(lv.color_hex(0x000000), 0)

slider = lv.slider(scrn)
slider.set_size(300, 50)
slider.center()

label = lv.label(scrn)
label.set_text('HELLO WORLD!')
label.align(lv.ALIGN.CENTER, 0, -50)
```

This is the command I used to build this:
```bash
/bin/python make.py esp32 BOARD=ESP32_GENERIC_S3 BOARD_VARIANT=SPIRAM_OCT --flash-size=8 DISPLAY=st7701 INDEV=ft6x36 EXPANDER=tca9535
```

It seems I had issues with the latest main branch, I can verify though that commit (b38f4d02a85daa85965650205eee3ecec5b7a9d5) worked with this though. After flashing and uploading the file, make sure you unplug for ~10s, before plugging it back in. If you soft reset too fast, it will also cause it not to work.

@kdschlosser Not sure if this helps with the latest changes, but the latest commit on main (7a50adbbaefea4400089da84fc361f0227869d4e) showed this:
```py
Touch Device ID: 0x00
Touch Chip ID: 0x00
Touch Device mode: 0x00
Touch Firmware ID: 0x00
Touch Release code: 0x00
Traceback (most recent call last):
  File "main.py", line 87, in <module>
  File "ft6x36.py", line 26, in __init__
  File "focaltech_touch.py", line 78, in __init__
RuntimeError: IC is not compatable with the FT6x36 driver
```

**Reply by kdschlosser** [reply]
I changed the I2C driver from being written in Python to being written in C. I had to do this to be able to get the I2C stuff to do a tear down properly. 

I am adding some logging to the driver so we will be able to see what is exactly happening with the data on the back end.

**Reply by kdschlosser** [reply]
I just pushed a commit that I am hoping will fix the touch.

**Reply by jbalague** [reply]
Thanks @grafail, now I get the same error you see. 

I saw that [esphome](https://esphome.io/components/touchscreen/ft5x06.html) supports the SenseCap Indicator and I gave it a try. it works flawlessly. I saw they are referencing another touch panel: the FT5x06. I tried to compile with INDEV=ft5x06 and changed your code this way:

```
# import ft6x36  # NOQA
import ft5x06
import i2c

i2c_bus = i2c.I2C.Bus(host=1, scl=40, sda=39, freq=5*100*1000, use_locks=False)
#touch_dev = i2c.I2C.Device(bus=i2c_bus, dev_id=72, reg_bits=ft6x36.BITS)
touch_dev = i2c.I2C.Device(bus=i2c_bus, dev_id=72, reg_bits=ft5x06.BITS)

#indev = ft6x36.FT6x36(touch_dev, startup_rotation=lv.DISPLAY_ROTATION._180)
indev = ft5x06.FT5x06(touch_dev, startup_rotation=lv.DISPLAY_ROTATION._180)
```

I got the same error.

Then I tried changing `dev_id` to `56` trying to match what I saw in esphome (`I2C_ADDR = 0x38`). It doesn't work either but error is different:

```
INITIALISING SCREEN
SPI WIRE DONE
RGB BUS DONE
BUFFER ALLOCATED
DISPLAY INITIALISED
Traceback (most recent call last):
  File "<stdin>", line 96, in <module>
  File "ft5x06.py", line 34, in __init__
  File "focaltech_touch.py", line 60, in __init__
  File "focaltech_touch.py", line 119, in _read_reg
OSError: [Errno 19] ENODEV
```

@kdschlosser,  maybe this helps you to narrow down the problem.

Thank you both for your support!

**Reply by grafail** [reply]
> Thanks @grafail, now I get the same error you see. 
> 
> I saw that [esphome](https://esphome.io/components/touchscreen/ft5x06.html) supports the SenseCap Indicator and I gave it a try. it works flawlessly. I saw they are referencing another touch panel: the FT5x06. I tried to compile with INDEV=ft5x06 and changed your code this way:
> 
> ```
> # import ft6x36  # NOQA
> import ft5x06
> import i2c
> 
> i2c_bus = i2c.I2C.Bus(host=1, scl=40, sda=39, freq=5*100*1000, use_locks=False)
> #touch_dev = i2c.I2C.Device(bus=i2c_bus, dev_id=72, reg_bits=ft6x36.BITS)
> touch_dev = i2c.I2C.Device(bus=i2c_bus, dev_id=72, reg_bits=ft5x06.BITS)
> 
> #indev = ft6x36.FT6x36(touch_dev, startup_rotation=lv.DISPLAY_ROTATION._180)
> indev = ft5x06.FT5x06(touch_dev, startup_rotation=lv.DISPLAY_ROTATION._180)
> ```
> 
> I got the same error.
> 
> Then I tried changing `dev_id` to `56` trying to match what I saw in esphome (`I2C_ADDR = 0x38`). It doesn't work either but error is different:
> 
> ```
> INITIALISING SCREEN
> SPI WIRE DONE
> RGB BUS DONE
> BUFFER ALLOCATED
> DISPLAY INITIALISED
> Traceback (most recent call last):
>   File "<stdin>", line 96, in <module>
>   File "ft5x06.py", line 34, in __init__
>   File "focaltech_touch.py", line 60, in __init__
>   File "focaltech_touch.py", line 119, in _read_reg
> OSError: [Errno 19] ENODEV
> ```
> 
> @kdschlosser,  maybe this helps you to narrow down the problem.
> 
> Thank you both for your support!

I saw that as well, but it seems it's wrong. If you try the code provided as is, it should work. Also make sure to checkout the commit I mentioned if main doesn't work.

**Reply by jbalague** [reply]
Yes, I tried with that commit and it worked. After the changes @kdschlosser made it did not. I'll try it again.

**Comment by spaetz** [comment]
Just got my hands on an indicator and would love to do some things in micropython on it. Is there a possibility to use the display from it by now? If not, no worries, I don't want to pressure anyone :-)

**Reply by kdschlosser** [reply]
I believe that he display driver is working. There are people using the driver I do know that. If there is something not working then open an issue (not a post in discussions) and I will address any issues you may have.

**Comment by dweber69** [comment]
@grafail I am hoping to use Micropython with my D1 as well. Were you able to get it to work in the end with all the changes discussed in this thread? Do you mind sharing your build command and a code sample? Thanks a lot in advance.
